### PR TITLE
feat: vamana

### PR DESCRIPTION
**Pull request by AveryQi115**
# Major Changes
- [x] Add in-memory vamana index into algorithms
- [x] Concurrently construct vamana index 
- [x] Fresh-vamana with insertion 
- [x] Reference filtered DiskANN to add filtered search for vamana index
  
# Next Plan - will develop in new PR: from vamana to diskANN
- [ ] Test basic vamana index performance and accuracy
- [ ] From vamana to diskANN (construct graphs by merging k-means clusters; searches using compressed vector; Beam Search)

### CODE REVIEW

**Comment by gaocegege** [comment]
Hi is it ready to review?

**Comment by AveryQi115** [comment]
> Hi is it ready to review?

Yes. I'm thinking about merging vamana first, and vamana can be used as an in memory index. And then I'll start a new pr to write disk based diskANN

**Comment by gaocegege** [comment]
Thanks for your contribution! :tada: :+1:

**Review by usamoi — APPROVED** [review]

File: src/algorithms/impls/vamana.rs [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+        let mut ids = (0..n).collect::<Vec<_>>();
+        ids.shuffle(&mut rng);
+
+        ids.into_par_iter()
+            .try_for_each(|id| self.search_and_prune_for_one_vertex(id, alpha, r, l))?;
+
+        Ok(())
+    }
+
+    #[warn(unused_assignments)]
```
[the remark this answers is not in the document]
Comment (AveryQi115) [reply src/algorithms/impls/vamana.rs:443]: Sure~

**Comment by VoVAllen** [comment]
@all-contributors please add @AveryQi115  for code contribution, Ideas

**Pull request merged**